flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/failover/flip1/SchedulingPipelinedRegionComputeUtil.java

```diff
@@ -0,0 +1,186 @@
+            }
+            mergedRegions.add(mergedRegion);
+        }
+
+        return mergedRegions;
+    }
+
+    private static List<List<Integer>> buildOutEdgesDesc(
+            final Map<SchedulingExecutionVertex, Set<SchedulingExecutionVertex>> vertexToRegion,
+            final List<Set<SchedulingExecutionVertex>> regionList,
+            final Function<ExecutionVertexID, ? extends SchedulingExecutionVertex>
+                    executionVertexRetriever) {
+
+        final Map<Set<SchedulingExecutionVertex>, Integer> regionIndices = new IdentityHashMap<>();
+        for (int i = 0; i < regionList.size(); i++) {
+            regionIndices.put(regionList.get(i), i);
+        }
+
+        final List<List<Integer>> outEdges = new ArrayList<>(regionList.size());
+        for (Set<SchedulingExecutionVertex> currentRegion : regionList) {
+            final List<Integer> currentRegionOutEdges = new ArrayList<>();
+            for (SchedulingExecutionVertex vertex : currentRegion) {
+                for (SchedulingResultPartition producedResult : vertex.getProducedResults()) {
+                    if (producedResult.getResultType().isPipelined()) {
+                        continue;
+                    }
+                    for (ConsumerVertexGroup consumerVertexGroup :
+                            producedResult.getConsumerVertexGroups()) {
+                        for (ExecutionVertexID consumerVertexId : consumerVertexGroup) {
+                            SchedulingExecutionVertex consumerVertex =
+                                    executionVertexRetriever.apply(consumerVertexId);
+                            // Since we are merging SchedulingPipelinedRegions inside one
+                            // LogicalPipelinedRegion, if any vertex of the ConsumerVertexGroup
+                            // doesn't belong to this LogicalPipelinedRegion, we can just skip the
+                            // remaining vertices in this group. This can decrease the computation
+                            // complexity.
+                            if (!vertexToRegion.containsKey(consumerVertex)) {
+                                break;
+                            }
+                            if (!currentRegion.contains(consumerVertex)) {
+                                currentRegionOutEdges.add(
+                                        regionIndices.get(vertexToRegion.get(consumerVertex)));
+                            }
+                        }
+                    }
+                }
+            }
+            outEdges.add(currentRegionOutEdges);
+        }
+
+        return outEdges;
+    }
+
+    private static Iterable<SchedulingResultPartition> getReconnectableResults(
+            SchedulingExecutionVertex vertex,
+            Function<IntermediateResultPartitionID, ? extends SchedulingResultPartition>
+                    resultPartitionRetriever) {
+        return () ->
```

Comment:
This method seems to be a bit too complex. Can it simply return a collection?

Reply:
I've tried to create an `ArrayList` and insert all satisfying results into the list. For 8k parallelism it will make the total time of `BuildTopologyBenchmark` increase from nearly 500ms to 2s. Thus, I think iterator would be better.

Reply:
In `LogicalPipelinedRegionComputUtil` the computation complexity is not disturbing us, thus, it returns a list.

Reply:
Ok. I did not expect this O(N) complexity operation to be that slow. It's then fine to me to take the iterator way.